Create pagefaultsnoop libbpf-tools

Rtoax opened the pull request:
Running like:

```bash
$ sudo ./pagefaultsnoop -t
TIME(s)  PID     COMM             TYPE  VM_FAULT        ADDR              
16:42:32 2467    smbd             FILE  .......N......  0x007f6157f86000
16:42:32 2467    smbd             FILE  .......N......  0x007f6157f98000
16:42:32 2912    virt-manager     WRITE ...W.....R....  0x007ff47d34c000
16:42:34 743     systemd-journal  FILE  ........L.....  0x007f02dae8c000
16:42:37 743     systemd-journal  FILE  ........L.....  0x007f02dae8d000
16:42:38 2912    virt-manager     WRITE ...W.....R....  0x007ff47d34c000
```

help info like:

```bash
$ sudo ./pagefaultsnoop --help
Usage: pagefaultsnoop [OPTION...]
Trace page fault.

USAGE: pagefaultsnoop [-h] [-t] [-x] [-p PID]

EXAMPLES:
    pagefaultsnoop             # trace page fault
    pagefaultsnoop -t          # include timestamps
    pagefaultsnoop -x          # ignore errors on output
    pagefaultsnoop -p 1216     # only trace PID 1216

VM_FAULT options are reported as:
  VM_FAULT_OOM                    O.............
  VM_FAULT_SIGBUS                 .S............
  VM_FAULT_MAJOR                  ..M...........
  VM_FAULT_WRITE                  ...W..........
  VM_FAULT_HWPOISON               ....h.........
  VM_FAULT_HWPOISON_LARGE         .....H........
  VM_FAULT_SIGSEGV                ......V.......
  VM_FAULT_NOPAGE                 .......N......
  VM_FAULT_LOCKED                 ........L.....
  VM_FAULT_RETRY                  .........R....
  VM_FAULT_FALLBACK               ..........F...
  VM_FAULT_DONE_COW               ...........C..
  VM_FAULT_NEEDDSYNC              ............s.
  VM_FAULT_HINDEX_MASK            .............m

  -p, --pid=PID              Process ID to trace
  -t, --timestamp            Include timestamp on output
  -v, --verbose              Verbose debug output
  -x, --failed               Include errors on output.
  -?, --help                 Give this help list
      --usage                Give a short usage message
  -V, --version              Print program version

```

Rtoax committed:
add pagefaultsnoop

Rtoax committed:
Create pagefaultsnoop.c

Rtoax committed:
Create pagefaultsnoop.bpf.c

Rtoax committed:
Create pagefaultsnoop.h

Rtoax committed:
Update pagefaultsnoop.c

brendangregg commented:
Thanks. Is there an example real world problem for this? I trace faults frequently for production issues, but I'm usually tracing the stack traces to see their cause. I can already do per-event traces, like this:

```
^Croot@bgregg-ThinkPad-T14-Gen-1:~> perf trace -e exceptions:page_fault_* -a
     0.000 Chrome_ChildIO/951829 exceptions:page_fault_user(address: 0xea801ba4fe0, ip: 139924376078569, error_code: 6)
     0.025 Chrome_ChildIO/951829 exceptions:page_fault_user(address: 0xea801ba550e, ip: 139924376079347, error_code: 6)
     0.509 Chrome_ChildIO/951829 exceptions:page_fault_user(address: 0x7f42a6aa9000, ip: 94066702965925, error_code: 6)
     1.289 chrome/952090 exceptions:page_fault_user(address: 0x7fa5b0b43000, ip: 94088014993372, error_code: 4)
```

And that already gives some info (address and error code). But I've never had a reason to use it.

I don't think we should add snoop tools for every tracepoint. Needs use cases. People should really be putting new tools (that aren't obviously useful) in their own personal repos (I have one myself), gathering some real world experience and use cases, and if proven useful bringing them here.

Rtoax committed:
Add -F FLAG option

Rtoax commented:
> ... if it had a flag option, like `-F flags`, then it might be easier to show how it solves use cases. For problem X, run `pagefaultsnoop -F RETRY`, or something like that. At the moment it's just dumping all events with flags and letting the end user figure it out. Being able to have a USAGE message that has a few targeted one-liners for real issues would help make the case.

Thanks for your reply, i just add `-F` option in https://github.com/iovisor/bcc/pull/3791/commits/16f0cde7441431b7763c05a25ae2acad54ac0c7a, usage:

```bash
$ sudo ./pagefaultsnoop -F LOCKED
Tracing Pagefault, Ctrl-C to end.
PID     COMM             TYPE  VM_FAULT        ADDR              
743     systemd-journal  FILE  ........L.....  0x007f02d9152000
743     systemd-journal  FILE  ........L.....  0x007f02d9153000
743     systemd-journal  FILE  ........L.....  0x007f02d9154000

$ sudo ./pagefaultsnoop -F RETRY 
Tracing Pagefault, Ctrl-C to end.
PID     COMM             TYPE  VM_FAULT        ADDR              
73163   pagefaultsnoop   WRITE ...W.....R....  0x007fbd09ca1000
73163   pagefaultsnoop   WRITE ...W.....R....  0x007fbd09bec000
73163   pagefaultsnoop   WRITE ...W.....R....  0x007fbd09c0e000
```

Rtoax committed:
set the -t and -F options at the same time error

brendangregg commented:
Thanks; so what's the story behind the tool? Every tool I create has some production issue behind it I couldn't solve with other tools.

Rtoax commented:
> perf trace -e exceptions:page_fault_* -a

Like you said, `perf trace -e exceptions:page_fault_* -a` can get pagefault easyly, but it may be miss some event like the following code:

```c
#include <malloc.h>

int main()
{
	int i;
	char *a = NULL;
	
	while (1) {
		a = malloc(40960);
		for (i = 0; i < 40960; i+=4096) {
			a[i] = 'a';
		}
	}
	return 0;
}
```

perf could LOST lots of event, see `LOST XXX events!`:

```
$ sudo perf trace -e exceptions:page_fault_* -a
[...]
LOST 11 events!
  3711.661 a.out/73827 exceptions:page_fault_user(address: 0x57060420, ip: 4195786, error_code: 6)
LOST 9 events!
  3711.706 a.out/73827 exceptions:page_fault_user(address: 0x5706a430, ip: 4195786, error_code: 6)
LOST 9 events!
  3711.756 a.out/73827 exceptions:page_fault_user(address: 0x57074440, ip: 4195786, error_code: 6)
LOST 10 events!
  3711.808 a.out/73827 exceptions:page_fault_user(address: 0x57089468, ip: 140589457648383, error_code: 6)
LOST 10 events!
  3711.857 a.out/73827 exceptions:page_fault_user(address: 0x5708a470, ip: 4195786, error_code: 6)
LOST 10 events!
  3711.907 a.out/73827 exceptions:page_fault_user(address: 0x57095480, ip: 4195786, error_code: 6)
LOST 10 events!
  3711.956 a.out/73827 exceptions:page_fault_user(address: 0x570a0490, ip: 4195786, error_code: 6)
LOST 10 events!
  3712.005 a.out/73827 exceptions:page_fault_user(address: 0x570ab4a0, ip: 4195786, error_code: 6)
LOST 8 events!
  3712.053 a.out/73827 exceptions:page_fault_user(address: 0x570b44b0, ip: 4195786, error_code: 6)
LOST 11 events!
  3712.103 a.out/73827 exceptions:page_fault_user(^Caddress: 0x570c04c0, ip: 4195786, error_code: 6)
```

however, bpf will not:

```
$ sudo ./pagefaultsnoop
Tracing Pagefault, Ctrl-C to end.
PID     COMM             TYPE  VM_FAULT        ADDR
[too many logs]
73892   a.out            ANON  ..............  0x000001b06a5000
73892   a.out            ANON  ..............  0x000001b06a6000
73892   a.out            ANON  ..............  0x000001b06a7000
73892   a.out            ANON  ..............  0x000001b06a8000
73892   a.out            ANON  ..............  0x000001b06a9000
73892   a.out            ANON  ..............  0x000001b06aa000
73892   a.out            ANON  ..............  0x000001b06ab000
73892   a.out            ANON  ..............  0x000001b06ac000
73892   a.out            ANON  ..............  0x000001b06b7000
73892   a.out            ANON  ..............  0x000001b06ae000
73892   a.out            ANON  ..............  0x000001b06af000
73892   a.out            ANON  ..............  0x000001b06b0000
73892   a.out            ANON  ..............  0x000001b06b1000
73892   a.out            ANON  ..............  0x000001b06b2000
73892   a.out            ANON  ..............  0x000001b06b3000
73892   a.out            ANON  ..............  0x000001b06b4000
73892   a.out            ANON  ..............  0x000001b06b5000
73892   a.out            ANON  ..............  0x000001b06b6000
73892   a.out            ANON  ..............  0x000001b06c1000
73892   a.out            ANON  ..............  0x000001b06b8000
73892   a.out            ANON  ..............  0x000001b06b9000
``` 

at the same time, easier to get VM_FAULT and VMA information.

Rtoax commented:
> Thanks; so what's the story behind the tool? Every tool I create has some production issue behind it I couldn't solve with other tools.



> Thanks; so what's the story behind the tool? Every tool I create has some production issue behind it I couldn't solve with other tools.

Like you said, `perf trace -e exceptions:page_fault_* -a` can get pagefault easyly, but it may be miss some event like the following code:

```c
#include <malloc.h>

int main()
{
	int i;
	char *a = NULL;
	
	while (1) {
		a = malloc(40960);
		for (i = 0; i < 40960; i+=4096) {
			a[i] = 'a';
		}
	}
	return 0;
}
```

perf could LOST lots of event, see `LOST XXX events!`:

```
$ sudo perf trace -e exceptions:page_fault_* -a
[...]
LOST 11 events!
  3711.661 a.out/73827 exceptions:page_fault_user(address: 0x57060420, ip: 4195786, error_code: 6)
LOST 9 events!
  3711.706 a.out/73827 exceptions:page_fault_user(address: 0x5706a430, ip: 4195786, error_code: 6)
LOST 9 events!
  3711.756 a.out/73827 exceptions:page_fault_user(address: 0x57074440, ip: 4195786, error_code: 6)
LOST 10 events!
  3711.808 a.out/73827 exceptions:page_fault_user(address: 0x57089468, ip: 140589457648383, error_code: 6)
LOST 10 events!
  3711.857 a.out/73827 exceptions:page_fault_user(address: 0x5708a470, ip: 4195786, error_code: 6)
LOST 10 events!
  3711.907 a.out/73827 exceptions:page_fault_user(address: 0x57095480, ip: 4195786, error_code: 6)
LOST 10 events!
  3711.956 a.out/73827 exceptions:page_fault_user(address: 0x570a0490, ip: 4195786, error_code: 6)
LOST 10 events!
  3712.005 a.out/73827 exceptions:page_fault_user(address: 0x570ab4a0, ip: 4195786, error_code: 6)
LOST 8 events!
  3712.053 a.out/73827 exceptions:page_fault_user(address: 0x570b44b0, ip: 4195786, error_code: 6)
LOST 11 events!
  3712.103 a.out/73827 exceptions:page_fault_user(^Caddress: 0x570c04c0, ip: 4195786, error_code: 6)
```

however, bpf will not:

```
$ sudo ./pagefaultsnoop
Tracing Pagefault, Ctrl-C to end.
PID     COMM             TYPE  VM_FAULT        ADDR
[too many logs]
73892   a.out            ANON  ..............  0x000001b06a5000
73892   a.out            ANON  ..............  0x000001b06a6000
73892   a.out            ANON  ..............  0x000001b06a7000
73892   a.out            ANON  ..............  0x000001b06a8000
73892   a.out            ANON  ..............  0x000001b06a9000
73892   a.out            ANON  ..............  0x000001b06aa000
73892   a.out            ANON  ..............  0x000001b06ab000
73892   a.out            ANON  ..............  0x000001b06ac000
73892   a.out            ANON  ..............  0x000001b06b7000
73892   a.out            ANON  ..............  0x000001b06ae000
73892   a.out            ANON  ..............  0x000001b06af000
73892   a.out            ANON  ..............  0x000001b06b0000
73892   a.out            ANON  ..............  0x000001b06b1000
73892   a.out            ANON  ..............  0x000001b06b2000
73892   a.out            ANON  ..............  0x000001b06b3000
73892   a.out            ANON  ..............  0x000001b06b4000
73892   a.out            ANON  ..............  0x000001b06b5000
73892   a.out            ANON  ..............  0x000001b06b6000
73892   a.out            ANON  ..............  0x000001b06c1000
73892   a.out            ANON  ..............  0x000001b06b8000
73892   a.out            ANON  ..............  0x000001b06b9000
``` 
You can see the ADDRESSes of pagefault are consecutive. At the same time, easier to get VM_FAULT and VMA information.

Rtoax committed:
update map value structure

Rtoax committed:
Update pagefaultsnoop.bpf.c

Rtoax committed:
Update pagefaultsnoop.c

Rtoax committed:
Update pagefaultsnoop.h

Rtoax committed:
Update pagefaultsnoop.bpf.c

Rtoax commented:
Commit https://github.com/iovisor/bcc/pull/3791/commits/8397a29868f43ce616201706656bf40eba092aef, https://github.com/iovisor/bcc/pull/3791/commits/1b46eecc73e5df2bb333b5c41dbc697210705165, https://github.com/iovisor/bcc/pull/3791/commits/b97d9ea47d51177ed3b6f4ad6aba2ec94abdf503, https://github.com/iovisor/bcc/pull/3791/commits/ebd62dc49d1b31a4603f01a7f6742c5d7822b702, https://github.com/iovisor/bcc/pull/3791/commits/817f60fa225d235dd890cb9e16315127a89fb0a9 add latency hist graph:

```bash
76983   a.out            FILE  ..M......R....  0x007f3877a49000 11356621
76983   a.out            FILE  .......N......  0x007f3877a49000 5264 
76983   a.out            FILE  ..M......R....  0x007f3877a4a000 16143776
76983   a.out            FILE  .......N......  0x007f3877a4a000 3382 
76983   a.out            FILE  ..M......R....  0x007f3877a49000 14073252
76983   a.out            FILE  .......N......  0x007f3877a49000 4065 
76983   a.out            FILE  ..M......R....  0x007f3877a4a000 7289273
76983   a.out            FILE  .......N......  0x007f3877a4a000 3084 
76983   a.out            FILE  ..M......R....  0x007f3877a49000 12485145
^CCleanup, output latency hists.
operation = 'FILE'
     nsecs               : count    distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 0        |                                        |
        16 -> 31         : 0        |                                        |
        32 -> 63         : 0        |                                        |
        64 -> 127        : 0        |                                        |
       128 -> 255        : 0        |                                        |
       256 -> 511        : 0        |                                        |
       512 -> 1023       : 0        |                                        |
      1024 -> 2047       : 0        |                                        |
      2048 -> 4095       : 17       |****************************            |
      4096 -> 8191       : 24       |****************************************|
      8192 -> 16383      : 7        |***********                             |
     16384 -> 32767      : 0        |                                        |
     32768 -> 65535      : 0        |                                        |
     65536 -> 131071     : 0        |                                        |
    131072 -> 262143     : 1        |*                                       |
    262144 -> 524287     : 7        |***********                             |
    524288 -> 1048575    : 2        |***                                     |
   1048576 -> 2097151    : 0        |                                        |
   2097152 -> 4194303    : 1        |*                                       |
   4194304 -> 8388607    : 11       |******************                      |
   8388608 -> 16777215   : 19       |*******************************         |
  16777216 -> 33554431   : 7        |***********                             |

operation = 'ANON'
     nsecs               : count    distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 0        |                                        |
        16 -> 31         : 0        |                                        |
        32 -> 63         : 0        |                                        |
        64 -> 127        : 0        |                                        |
       128 -> 255        : 0        |                                        |
       256 -> 511        : 0        |                                        |
       512 -> 1023       : 0        |                                        |
      1024 -> 2047       : 1082384  |****************************************|
      2048 -> 4095       : 621867   |**********************                  |
      4096 -> 8191       : 139246   |*****                                   |
      8192 -> 16383      : 26195    |                                        |
     16384 -> 32767      : 3029     |                                        |
     32768 -> 65535      : 64       |                                        |
     65536 -> 131071     : 4        |                                        |
    131072 -> 262143     : 53       |                                        |
    262144 -> 524287     : 73       |                                        |
    524288 -> 1048575    : 37       |                                        |
   1048576 -> 2097151    : 53       |                                        |
   2097152 -> 4194303    : 21       |                                        |
   4194304 -> 8388607    : 8        |                                        |
   8388608 -> 16777215   : 17       |                                        |
  16777216 -> 33554431   : 1        |                                        |
```

netedwardwu commented:
Hi,

I have the same question...what is the real world case needs this tool?
Why do we need this tool?
Thank you

Rtoax commented:
> Hi,
> 
> I have the same question...what is the real world case needs this tool? Why do we need this tool? Thank you

Sometimes, too much page fault cause bad performance, CAN use this tool checkout all type of Page Fault, 

* When too much SWAP page fault, decide whether to close SWAP based on a specific application;
* When too much ANON page fault, decide whether need to call `mlock()` or `mlockall()`, mlock can be a disaster for small memory system and cannot be used in all situations;
* When too much FILE page fault in computationally intensive program, this can be used as a reference for program optimization;
* When too much NUMA page fault, decide whether to turn off NUMA according to the memory model, As far as I know, Tencent turns on NUMA by default, and Alibaba Cloud turns off NUMA by default. I believe this tool will give a reference.

chenhengqi commented:
Why not use tracepoint in this tool ?

Rtoax commented:
> Why not use tracepoint in this tool ?

Cause don't have tracepoint in `do_anonymous_page`, `do_swap_page`, `do_fault`, `do_numa_page` and `do_wp_page`:

```bash
$ sudo bpftrace -l | \
   grep -e do_anonymous_page -e do_fault -e do_swap_page -e do_numa_page -e do_wp_page
kfunc:__do_fault
kfunc:do_anonymous_page
kfunc:do_fault
kfunc:do_numa_page
kfunc:do_swap_page
kfunc:do_wp_page
kprobe:__do_fault
kprobe:do_anonymous_page
kprobe:do_fault
kprobe:do_numa_page
kprobe:do_swap_page
kprobe:do_wp_page
```

There are `tracepoint:exceptions:page_fault_kernel` and `tracepoint:exceptions:page_fault_user`, but it's unuseful for this tool, page fault call stack like:

```
handle_pte_fault
    do_anonymous_page
    do_fault 
         do_read_fault
         do_cow_fault
         do_shared_fault
    do_swap_page 
         lookup_swap_cache
         alloc_page_vma
    do_numa_page
         vm_normal_page
    do_wp_page
         vm_normal_page
         wp_pfn_shared
         wp_page_copy
         wp_page_reuse
         wp_page_shared
```

Further more, maybe add some meaningful sub-type of page fault, like `do_fault` can continue to be divided into `do_read_fault`, `do_cow_fault`, `do_shared_fault`. However, So far, there are no tracing method for them:

```bash
$ sudo bpftrace -l | grep do_cow_fault
[get nothing]
```

Rtoax committed:
kprobe:do_fault to kprobe:__do_fault

Different kernel version, `do_fault` kprobe are different:

```bash
$ sudo bpftrace -l | grep do_fault
kfunc:__do_fault
kfunc:do_fault
kprobe:__do_fault
kprobe:do_fault
$ uname -r
5.14.0
```

in `5.15`:

```bash
```bash
$ sudo bpftrace -l | grep do_fault
kfunc:__do_fault
kprobe:__do_fault
$ uname -r
5.15.2
```
```

Rtoax commented:
[commit 9bfca5e](https://github.com/iovisor/bcc/pull/3791/commits/9bfca5e5bbf88359b2698ddf29a6ce3e5fa9e691) fix:

```bash
$ uname -r
5.15.2
$ sudo ./pagefaultsnoop 
libbpf: kprobe perf_event_open() failed: No such file or directory
libbpf: prog 'do_fault_entry': failed to create kprobe 'do_fault+0x0' perf event: No such file or directory
libbpf: failed to auto-attach program 'do_fault_entry': -2
failed to attach BPF programs: -2
```

Rtoax committed:
add pagefaultsnoop to .gitignore

Rtoax commented:
I just find Fedora35 miss all kprobe of `do_xxx_page`:

```bash
$ sudo bpftrace -l | grep do_*_page
kfunc:do_page_add_anon_rmap
kfunc:do_page_cache_ra
kfunc:do_page_mkwrite
kfunc:do_pages_stat
kprobe:do_page_add_anon_rmap
kprobe:do_page_cache_ra
kprobe:do_page_mkwrite
kprobe:do_pages_stat
$ uname -r
5.15.8-200.fc35.x86_64
```

These kprobes maybe relate to `OS relase version`, the `pagefaultsnoop` can use in CentOS
